dogstatsd-client/src/lib.rs

```diff
@@ -0,0 +1,331 @@
+use serde::{Deserialize, Serialize};
+use std::fmt::Debug;
+use tracing::{debug, error, info};
+
+use anyhow::anyhow;
+use cadence::prelude::*;
+#[cfg(unix)]
+use cadence::UnixMetricSink;
+use cadence::{Metric, MetricBuilder, QueuingMetricSink, StatsdClient, UdpMetricSink};
+#[cfg(unix)]
+use ddcommon::connector::uds::socket_path_from_uri;
+use std::net::{ToSocketAddrs, UdpSocket};
+#[cfg(unix)]
+use std::os::unix::net::UnixDatagram;
+
+// Queue with a maximum capacity of 32K elements
+const QUEUE_SIZE: usize = 32 * 1024;
+
+/// The `DogStatsDActionOwned` enum gathers the metric types that can be sent to the DogStatsD
+/// server. This type takes ownership of the relevant data to support the sidecar better.
+/// For documentation on the dogstatsd metric types: https://docs.datadoghq.com/metrics/types/?tab=count#metric-types
+///
+/// Originally I attempted to combine this type with `DogStatsDAction` but this GREATLY complicates
+/// the types to the point of insanity. I was unable to come up with a satisfactory approach that
+/// allows both the data-pipeline and sidecar crates to use the same type. If a future rustacean
+/// wants to take a stab and open a PR please do so!
+#[derive(Debug, Serialize, Deserialize)]
+pub enum DogStatsDActionOwned {
+    #[allow(missing_docs)]
+    Count(String, i64, Vec<Tag>),
+    #[allow(missing_docs)]
+    Distribution(String, f64, Vec<Tag>),
+    #[allow(missing_docs)]
+    Gauge(String, f64, Vec<Tag>),
+    #[allow(missing_docs)]
+    Histogram(String, f64, Vec<Tag>),
+    /// Cadence only support i64 type as value
+    /// but Golang implementation uses string (https://github.com/DataDog/datadog-go/blob/331d24832f7eac97b091efd696278fe2c4192b29/statsd/statsd.go#L230)
+    /// and PHP implementation uses float or string (https://github.com/DataDog/php-datadogstatsd/blob/0efdd1c38f6d3dd407efbb899ad1fd2e5cd18085/src/DogStatsd.php#L251)
+    Set(String, i64, Vec<Tag>),
+}
+
+/// The `DogStatsDAction` enum gathers the metric types that can be sent to the DogStatsD server.
+#[derive(Debug, Serialize, Deserialize)]
+pub enum DogStatsDAction<'a, T: AsRef<str>, V: IntoIterator<Item = &'a Tag>> {
+    // TODO: instead of AsRef<str> we can accept a marker Trait that users of this crate implement
+    #[allow(missing_docs)]
+    Count(T, i64, V),
+    #[allow(missing_docs)]
+    Distribution(T, f64, V),
+    #[allow(missing_docs)]
+    Gauge(T, f64, V),
+    #[allow(missing_docs)]
+    Histogram(T, f64, V),
+    /// Cadence only support i64 type as value
+    /// but Golang implementation uses string (https://github.com/DataDog/datadog-go/blob/331d24832f7eac97b091efd696278fe2c4192b29/statsd/statsd.go#L230)
+    /// and PHP implementation uses float or string (https://github.com/DataDog/php-datadogstatsd/blob/0efdd1c38f6d3dd407efbb899ad1fd2e5cd18085/src/DogStatsd.php#L251)
+    Set(T, i64, V),
+}
+
+/// A dogstatsd-client that flushes stats to a given endpoint. Use `new_flusher` to build one.
+pub struct Flusher {
```

Comment:
Nit: Should we take the opportunity to rename it to `Client` or something similar?